### ISSUE TITLE
"The value '' cannot be parsed as the type 'Int32'

### ISSUE THREAD

**Issue by ranjan-web** [XiaoFaye/WooCommerce.NET#518]
WooCommerce.NET, Version=0.7.8.0
WordPress 5.5
WooCommerce version: 4.3.2

I'm trying to batch update Category and following error occurred.  Can help on this issue?
Please note that, I have used **jsonDeserializeFilter**: trimstr in request.

**Inner Exception** : {"The value '' cannot be parsed as the type 'Int32'."}
**Message** : There was an error deserializing the object of type WooCommerceNET.Base.BatchObject`1[[WooCommerceNET.WooCommerce.v3.ProductCategory, WooCommerce.NET, Version=0.7.8.0, Culture=neutral, PublicKeyToken=ff41d0b6c937438d]]. The value '' cannot be parsed as the type 'Int32'.

Following is my code

RestAPI rest = new RestAPI("https://my-domain.com/sync/wp-json/wc/v3/",
    "ck_a445e37faf3015xxxxda28a5b36fa04", "cs_d3b7a0f8b5fd0d1xxxx08113bde52996ed",
    jsonDeserializeFilter: trimstr);
            WCObject wc = null;
            wc = new WCObject(rest);

            //Batch add/update/delete
            ProductCategoryBatch cb = new ProductCategoryBatch();

            List<ProductCategory> create = new List<ProductCategory>();
            create.Add(new ProductCategory()
            {
                name = "test_category"
            });

            List<ProductCategory> update = new List<ProductCategory>();
            update.Add(new ProductCategory()
            {
                id = 215,
                description = "description updated"
            });

            List<int> delete = new List<int>() { 214 };
            cb.create = create;
            cb.update = update;
            cb.delete = delete;

            var c = wc.Category.UpdateRange(cb).Result;

![image](https://user-images.githubusercontent.com/71118925/92881696-d820fb80-f42c-11ea-9a11-71bd0e6c917e.png)

**Comment by ioandev** [comment]
Hi @ranjan-web , i'm having the same issue.

  | Name | Value | Type
-- | -- | -- | --
◢ | batch | {WooCommerceNET.Base.BatchObject<WooCommerceNET.WooCommerce.v3.Variation>} | WooCommerceNET.Base.BatchObject<WooCommerceNET.WooCommerce.v3.Variation>
  | ▶ create | Count = 132 | System.Collections.Generic.List<WooCommerceNET.WooCommerce.v3.Variation>
  | ◢ delete | Count = 10 | System.Collections.Generic.List<int>
  | [0] | 3193 | int
  | [1] | 3192 | int
  | [2] | 3191 | int
  | [3] | 3190 | int
  | [4] | 3189 | int
  | [5] | 3188 | int
  | [6] | 3187 | int
  | [7] | 3186 | int
  | [8] | 3185 | int
  | [9] | 3184 | int

```
System.Runtime.Serialization.SerializationException: There was an error deserializing the object of type WooCommerceNET.Base.BatchObject`1[[WooCommerceNET.WooCommerce.v3.Variation, WooCommerce.NET, Version=0.8.2.0, Culture=neutral, PublicKeyToken=ff41d0b6c937438d]]. The value '' cannot be parsed as the type 'Int32'.
 ---> System.Xml.XmlException: The value '' cannot be parsed as the type 'Int32'.
 ---> System.FormatException: Input string was not in a correct format.
   at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
   at System.Number.ParseInt32(ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
   at System.Runtime.Serialization.Json.XmlJsonReader.ParseInt(String value, NumberStyles style)
   --- End of inner exception stack trace ---
   at System.Runtime.Serialization.Json.XmlJsonReader.ParseInt(String value, NumberStyles style)
   at System.Xml.XmlDictionaryReader.ReadElementContentAsInt()
   at ReadArrayOfintFromJson(XmlReaderDelegator , XmlObjectSerializerReadContextComplexJson , XmlDictionaryString , XmlDictionaryString , CollectionDataContract )
   at System.Runtime.Serialization.Json.JsonCollectionDataContract.ReadJsonValueCore(XmlReaderDelegator jsonReader, XmlObjectSerializerReadContextComplexJson context)
   at System.Runtime.Serialization.Json.JsonDataContract.ReadJsonValue(XmlReaderDelegator jsonReader, XmlObjectSerializerReadContextComplexJson context)
   at System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(XmlReaderDelegator reader, String name, String ns, Type declaredType, DataContract& dataContract)
   at System.Runtime.Serialization.XmlObjectSerializerReadContextComplex.InternalDeserialize(XmlReaderDelegator xmlReader, Int32 declaredTypeID, RuntimeTypeHandle declaredTypeHandle, String name, String ns)
   at ReadBatchObjectOfVariationDXupGQmZFromJson(XmlReaderDelegator , XmlObjectSerializerReadContextComplexJson , XmlDictionaryString , XmlDictionaryString[] )
   at System.Runtime.Serialization.Json.JsonClassDataContract.ReadJsonValueCore(XmlReaderDelegator jsonReader, XmlObjectSerializerReadContextComplexJson context)
   at System.Runtime.Serialization.Json.JsonDataContract.ReadJsonValue(XmlReaderDelegator jsonReader, XmlObjectSerializerReadContextComplexJson context)
   at System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(XmlReaderDelegator reader, String name, String ns, Type declaredType, DataContract& dataContract)
   at System.Runtime.Serialization.XmlObjectSerializerReadContextComplex.InternalDeserialize(XmlReaderDelegator xmlReader, Type declaredType, DataContract dataContract, String name, String ns)
   at System.Runtime.Serialization.Json.DataContractJsonSerializerImpl.InternalReadObject(XmlReaderDelegator xmlReader, Boolean verifyObjectName)
   at System.Runtime.Serialization.XmlObjectSerializer.InternalReadObject(XmlReaderDelegator reader, Boolean verifyObjectName, DataContractResolver dataContractResolver)
   at System.Runtime.Serialization.XmlObjectSerializer.ReadObjectHandleExceptions(XmlReaderDelegator reader, Boolean verifyObjectName, DataContractResolver dataContractResolver)
   --- End of inner exception stack trace ---
   at WooCommerceNET.RestAPI.DeserializeJSon[T](String jsonString)
   at WooCommerceNET.Base.WCSubItem`1.UpdateRange(Int32 parentId, BatchObject`1 items, Dictionary`2 parms)
   at NextContact.NextWooCommerce.UpdateAll(List`1 sourceProducts, List`1 destinationProducts) in G:\Projects\DynamicPriceCalculator\NextContact\NextWooCommerce.cs:line 112
```

**Comment by XiaoFaye** [comment]
Set RestAPI.DEBUG to true, you should see the full json string in exception message, then  you can verify which part of the json causing this issue.

**Comment by ranjan-web** [comment]
Thank you, Will try that

**Comment by vankyy26** [comment]
I get similar exception with latest update 0.8.2. That's the code:

```
Dictionary<string, string> dic = new Dictionary<string, string>();
dic.Add("sku", BaseProductSku);
dic.Add("per_page", "1");

var result = await wc.Product.GetAll(dic);
```

![image](https://user-images.githubusercontent.com/10697327/94868935-d3a89b00-044c-11eb-8a0c-cb8e365aafa7.png)

Woocommerce version 4.5.2
WordPress version 5.5.1

**Comment by vankyy26** [comment]
After further diffing I see what's the problem. Now the API returns the variations as full objects, not List<int>. I enabled Debug mode and this is the JSON I receive from WooApi: 

![image](https://user-images.githubusercontent.com/10697327/94871384-85969600-0452-11eb-849d-cf72d3574d42.png)

0.8.2 expects this: 

![image](https://user-images.githubusercontent.com/10697327/94871427-9ba45680-0452-11eb-87f4-092076fb20a6.png)

As it's important for me to fix this now, I'll checkout the source locally. I can send you a PR with a fix.

**Comment by XiaoFaye** [comment]
@vankyy26 So you mean in the new version of WooCommerce, the variations returns as list of variation object instead of list of int?

I think upload of Media only supports local image file upload, not image url?

**Comment by vankyy26** [comment]
> @vankyy26 So you mean in the new version of WooCommerce, the variations returns as list of variation object instead of list of int?
> 
> I think upload of Media only supports local image file upload, not image url?

Hi and sorry for the late response. Yes, I'm uploading a local file. The workaround is to upload the image via FTP and in the product creation use the link of the uploaded image, ugly but works. 

About the variations, looks like yes, they are objects now, as the existing dto do not deserialize properly I switched the type of the list to object. This way it will skip it. There is new version of Woocommerce and I have to try but I suppose it will be the same. This is how I changed the variations list.

![image](https://user-images.githubusercontent.com/10697327/98445054-34975300-211e-11eb-933d-9601695cdc93.png)

I'm worried to update back to the NuGet package at the moment and I'll keep using my local version until you confirm it works normally. I use it for a POS app and it's important to work.
